New Agent: wiregrid_tiltsensor agent

d-hoshino2626 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
This pull request contains agent and several driver codes to handle wiregrid tilt sensors.
There are two types of tilt sensors, so two drivers are contained.

New Agent: wiregrid_tiltsensor
This agent records the data sent by tilt sensor via a serial-to-ethernet converter.
New drivers:
dwl.py: The driver to communicate DWL5000XY tilt sensor
sherborne.py: The driver to communicate with sherborne tilt sensor

## Motivation and Context
Add this agent to the main branch.

## How Has This Been Tested?
This agent has been tested in a test bench of wire grid at Kyoto University only with DWL tilt sensor via USB.
The test with sherborne tilt sensor has not performed yet.
Also, the test with a serial-to-ethernet converter (the same condition as at site) has not performed yet.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

BrianJKoopman commented on socs/agents/wiregrid_tiltsensor/drivers/dwl.py:
```diff
@@ -0,0 +1,235 @@
+    # ***** Helper Methods *****
+
+    def __conn(self, rtu_port=None, tcp_ip=None, tcp_port=None, timeout=None):
+        """
+        Connect to the PMX module
+
+        Args:
+        rtu_port (str): Serial RTU port
+        tcp_ip (str): TCP IP address
+        tcp_port (int): TCP port
+        """
```
I thought I recognized this code. This clearly reuses bits of the PMX module. That's fine, though I one point I worked on abstracting some of this and reducing duplicated code (which I would still advocate for.)

The thing I am wondering here though is if the whole `rtu_port` structure makes sense? For a few reasons:
* I think the RTU in this context refers to [Modbus RTU](https://en.wikipedia.org/wiki/Modbus#Modbus_RTU), and correct me if I'm wrong, but I don't believe this device supports modbus.
* It seems a Serial to Ethernet converter is always used with the agent. I believe here the RTU port structure is trying to select between connecting directly to the device via Serial (to use the Modbus RTU interface) or over TCP using the Serial to Ethernet converter.

If we don't support direct serial connection to the device, this `rtu_port` is a bunch of cruft for no functionality and should be removed. If there's some reason to support direct serial connection outside of the Agent, do let me know though.

d-hoshino2626 replied:
I concluded the direct serial connection will not be used via ocs-agent, so I removed all rtu option in this agent and drivers.

BrianJKoopman reviewed (changes requested):
Thanks for the PR! Sorry again for the delay in the review. Here are some comments mostly focusing on the Agent. I might have some more to say about the driver code, but these should get you started. I likely won't have time to add any more thoughts until next week.

I'm also going to push a commit that adds an outline for the docs page so I could see the nicely formatted docstrings you included for the agent's operations. (Thanks!) Can you fill in some of the detail there?

d-hoshino2626 commented on socs/agents/wiregrid_tiltsensor/agent.py:
```diff
@@ -0,0 +1,230 @@
+                if self.type_tiltsensor == 'sherborne':
+                    msg, temperatures = self.tiltsensor.get_temp()
+                else:
+                    pass
```
I removed all extra `else: pass`.

d-hoshino2626 commented on socs/agents/wiregrid_tiltsensor/drivers/dwl.py:
```diff
@@ -0,0 +1,235 @@
+            print(
+                "Disconnecting from TCP IP %s at port %d"
+                % (self.tcp_ip, self.tcp_port))
+            pass
```
I removed extra `pass`.

d-hoshino2626 commented on socs/agents/wiregrid_tiltsensor/agent.py:
```diff
@@ -0,0 +1,230 @@
+                    .format(self.lock.job))
+                return False, 'Could not acquire lock.'
+
+            session.set_status('running')
```
I removed this line.

d-hoshino2626 commented on socs/agents/wiregrid_tiltsensor/agent.py:
```diff
@@ -0,0 +1,230 @@
+            There are twp types of tilt sensor,
+            and this argument is used for specifying
+            to communicate with whichtilt sensor.
+            This argument should be 'DWL' or 'sherborne'.
```
I modified `tiltsensor_ip` -> `ip`, `tiltsensor_port` -> `port` and `type_tiltsensor` -> `sensor_type`.

d-hoshino2626 commented on socs/agents/wiregrid_tiltsensor/agent.py:
```diff
@@ -0,0 +1,230 @@
+        tiltsensor = Sherborne(tcp_ip=tiltsensor_ip, tcp_port=tiltsensor_port, reset_boot=False, timeout=0.5, verbose=0)
+    else:
+        raise ('Invalid tiltsensor type')
+    return tiltsensor
```
I added the lines you commented in `socs/agents/wiregrid_tiltsensor/drivers/__init__.py` and `from socs.agents.wiregrid_tiltsensor.drivers import connect` in `agent.py`

d-hoshino2626 commented on socs/agents/wiregrid_tiltsensor/agent.py:
```diff
@@ -0,0 +1,230 @@
+    pgroup = parser_in.add_argument_group('Agent Options')
+    pgroup.add_argument('--tiltsensor_ip')
+    pgroup.add_argument('--tiltsensor_port')
+    pgroup.add_argument('--type_tiltsensor',
```
I chenged these arguments to `--ip-address`, `--port` and `--sensor-type`.

d-hoshino2626 commented on socs/agents/wiregrid_tiltsensor/agent.py:
```diff
@@ -0,0 +1,230 @@
+                    f.write(str(last_release) + '\n')
+                    pass
+                """
+                # self.log.info(f"sens_data:{sens_data}")
```
I removed these lines.

d-hoshino2626 commented on socs/agents/wiregrid_tiltsensor/agent.py:
```diff
@@ -0,0 +1,230 @@
+                return False
+
+            if self.type_tiltsensor != 'sherborne':
+                raise ("This type of tiltsensor cannot reset.")
```
I changed as you say.

d-hoshino2626 commented on socs/agents/wiregrid_tiltsensor/agent.py:
```diff
@@ -0,0 +1,230 @@
+                    'timestamp': current_time
+                }
+
+                time.sleep(0.5)  # DAQ interval
```
Thanks for your useful information! Now I use Pacemaker instead of time.sleep().

d-hoshino2626 commented on socs/agents/wiregrid_tiltsensor/agent.py:
```diff
@@ -0,0 +1,230 @@
+    def reset(self, session, params=None):
+        """reset()
+
+        **Task** - Reset the tiltsensor if the type of tiltsensor is sherborne.
```
As you say, just restart the device.

d-hoshino2626 commented on socs/agents/wiregrid_tiltsensor/drivers/dwl.py:
```diff
@@ -0,0 +1,235 @@
+
+            if len(read) >= SIZE:
+                break
+            pass
```
I modified this method. In new code, I removed the loop. I expect to recieve a 12 byte response and check the initial 2 bytes match the header. If matched, calculate the output angle, and if not matched, return -999.

d-hoshino2626 commented:
Thank you for reviewing, Brian! Your review is very helpful for me. I modified some codes according to your review and I added docs about this agent. If you have any question or request, please let me know freely.

BrianJKoopman commented on socs/agents/wiregrid_tiltsensor/drivers/dwl.py:
```diff
@@ -0,0 +1,146 @@
+        # write and read serial
+        self.ser.write(command)
+        read_hex = self.ser.read(SIZE)
+        if self.verbose > 0:
+            print(f'{read_hex}')
```
This gets printed again anyway on line 53. Any reason to keep this? If not, remove it.

d-hoshino2626 replied:
I removed it. (022b0cc)

BrianJKoopman commented on socs/agents/wiregrid_tiltsensor/drivers/dwl.py:
```diff
@@ -0,0 +1,146 @@
+            readInt = []
+            val = ()
+            for c in read:
+                readInt.append((int)(c, 16))
```
Does `self.ser.read(SIZE)` always return something? It seems like it would be possible to get nothing back from the read and end up with an empty `readInt`, which would cause issues down on lines 67, 78, 89, etc. I'd suggest adding some check that `read` is the expected size.

d-hoshino2626 replied:
Your concern is very reasonable. I modified it to return -999 or (-999, -999) if `read` does not have the expected size. (022b0cc)

BrianJKoopman reviewed (changes requested):
Thanks for making the changes in response to the last review! And sorry this has taken so long to get back to. I have a few quick comments on the current state, and I pushed a few small commits to address a couple of small things I noticed while re-reviewing. Let me know if you have any questions.

d-hoshino2626 commented on docs/agents/wiregrid_tiltsensor.rst:
```diff
@@ -0,0 +1,85 @@
+ - reset():
+   Reset the sherborne tilt sensor.
+
+
```
I removed it in commit a00180b.

d-hoshino2626 commented on docs/agents/wiregrid_tiltsensor.rst:
```diff
@@ -0,0 +1,85 @@
+- tiltsensor-port is an asigned port for the tilt sensor.
+  (The converter has four D-sub ports to control
+  multiple devices connected via serial communication.
+  Communicating device is determined by the ethernet port number of the converter.)
```
I clearly added that, the 4 D-sub ports are 23, 26, 29, 32. (a00180b)

d-hoshino2626 commented on socs/agents/wiregrid_tiltsensor/agent.py:
```diff
@@ -0,0 +1,208 @@
+                # Execute reset()
+                self.tiltsensor.reset()
+                # Store the timestamp when reset is performed in session.data
+                session.data = {'reset': True,
+                                'timestamp': time.time()}
```
I added the docstring. (49b129b)

d-hoshino2626 commented:
Thank you for reviewing and helpful commits @BrianJKoopman! I fixed problems according to your comment. And I also found a problem in `agent.py` so I modified it in commit 393ea2c.

BrianJKoopman reviewed (approved):
Thanks for addressing all the comments, this is ready to merge. I just pushed one final commit to add the agent to the plugins list, I'll merge once checks complete.
